File: p1/p1.cpp
class Foo {
private:
    int currentTurn = 1;
    mutex m;
    condition_variable cv;

    void run_on_turn(function<void()> runnable, int turn) {
        {
            unique_lock<mutex> lock(m);
            cv.wait(lock, [this, turn]{ return currentTurn == turn; });
        }
        runnable();
        {
            unique_lock<mutex> lock(m);
            currentTurn++;
        }
        cv.notify_all();
    }

public:
    Foo() {
        
    }

    void first(function<void()> printFirst) {
        run_on_turn(printFirst, 1);
    }

    void second(function<void()> printSecond) {
        run_on_turn(printSecond, 2);
    }

    void third(function<void()> printThird) {
        run_on_turn(printThird, 3);
    }
};

File: p1/p1.java
class Foo {

    int nextIteration = 1;   

    public Foo() {
        
    }

    private void runOnTurn(Runnable runnable, int turn) throws InterruptedException {
        synchronized(this) {
            while (nextIteration < turn) {
                this.wait();
            }
            if (nextIteration != turn) {   throw new IllegalStateException();  }
        }
        runnable.run();
        synchronized(this) {
            nextIteration++;
            this.notifyAll();
        }
    }

    public void first(Runnable printFirst) throws InterruptedException {
        runOnTurn(printFirst, 1);
    }

    public void second(Runnable printSecond) throws InterruptedException {
        runOnTurn(printSecond, 2);
    }

    public void third(Runnable printThird) throws InterruptedException {
        runOnTurn(printThird, 3);
    }
}